### DISCUSSION TITLE
How to display dynamic ordered content?

### DISCUSSION THREAD

**Discussion by dpkristensen** [Textualize/textual#3319] [General]
I'm porting an app I wrote in tkinter and running into some issues.  Neither Tree nor ListView seem to support inserting items into the middle of the list or sorting in general.  I am communicating with a device and receiving information at runtime, so it is not possible to know this at the time "compose" is run.

I thought about creating a Vertical Scroll container, then adding a collapsible Container of labels.  There's a couple issues though.  Adding to the containers use a compose function (which seems risky to call this outside of `compose()`), and is always to the end; I'd rather not add it to the end and then move it, why not just specify the index when I am inserting?

I suppose the only control capable of doing this is the Data Table?  Or just format the text as a blob and put it into a non-editable text box?

When inserting things into the list, the container has to be redrawn, and it may affect things like scroll bars, currently selected item, etc...  So I realize it's not exactly trivial; just wondering if there was something I'm missing at a higher level.

**Comment by davep** [comment]
> Adding to the containers use a compose function (which seems risky to call this outside of compose()), and is always to the end; I'd rather not add it to the end and then move it, why not just specify the index when I am inserting?

You might want to look back over the Textual tutorial, especially at the part about [dynamic widgets](https://textual.textualize.io/tutorial/#dynamic-widgets); `compose` isn't the only way to add widgets into a containing widget; and [the method you'd use for that lets you specify where to add them](https://textual.textualize.io/api/widget/#textual.widget.Widget.mount). You can even [move them later](https://textual.textualize.io/api/widget/#textual.widget.Widget.move_child).

**Comment by dpkristensen** [comment]
Ok, I think the terminology of "mount" is confusing; I haven't seen this in any other UI framework I've worked with.  There are several methods using "child" in the terminology, I think it would be helpful to have a public method add_child alias for the mount() method.

Of course if Tree and List had a way to add an entry by index, that would be helpful as well. 😄
